src/client/executor.rs

```diff
@@ -309,167 +285,152 @@ impl Client {
-                // locking, so users need to disable retryable writes if using mmapv1.
-                if let ErrorKind::Command(ref mut command_error) = *err.kind {
-                    if command_error.code == 20
-                        && command_error.message.starts_with("Transaction numbers")
-                    {
-                        command_error.message = "This MongoDB deployment does not support \
-                                                 retryable writes. Please add retryWrites=false \
-                                                 to your connection string."
-                            .to_string();
+                    let op_retry = match self.get_op_retryability(&op, &session) {
+                        Retryability::Read => err.is_read_retryable(),
+                        Retryability::Write => err.is_write_retryable(),
+                        _ => false,
+                    };
+                    if err.is_pool_cleared() || op_retry {
+                        retry = Some(ExecutionRetry {
+                            prior_txn_number: None,
+                            first_error: err,
+                        });
+                        continue;
+                    } else {
+                        return Err(err);
                     }
                 }
+            };
 
-                self.inner
-                    .topology
-                    .handle_application_error(
-                        server.address.clone(),
-                        err.clone(),
-                        HandshakePhase::after_completion(&conn),
-                    )
-                    .await;
-                // release the connection to be processed by the connection pool
-                drop(conn);
-                // release the selected server to decrement its operation count
-                drop(server);
-
-                if retryability == Retryability::Read && err.is_read_retryable()
-                    || retryability == Retryability::Write && err.is_write_retryable()
-                {
-                    self.execute_retry(&mut op, &mut session, txn_number, err)
-                        .await
-                } else {
-                    Err(err)
-                }
+            if session.is_none() {
+                implicit_session = self.start_implicit_session(&op).await?;
+                session = implicit_session.as_mut();
             }
-        }
-    }
 
-    async fn execute_retry<T: Operation>(
-        &self,
-        op: &mut T,
-        session: &mut Option<&mut ClientSession>,
-        prior_txn_number: Option<i64>,
-        first_error: Error,
-    ) -> Result<ExecutionOutput<T>> {
-        op.update_for_retry();
-
-        let server = match self.select_server(op.selection_criteria()).await {
-            Ok(server) => server,
-            Err(_) => {
-                return Err(first_error);
+            let retryability = self.get_retryability(&conn, &op, &session)?;
```

Comment:
Is there any chance this could fail during a retry? If so, I think we should return the original error rather than the one that occurs in this call.

Reply:
The only error this can return is an internal one indicating an impossible/invalid connection state, which I think would be valuable to expose if it does happen :)